app/controllers/api/alert_definitions_controller.rb

```diff
@@ -31,5 +28,10 @@ def edit_resource(type, id = nil, data = {})
     def set_additional_attributes
       @additional_attributes = %w(expression)
     end
+
+    def update_miq_expression(data)
+      data["miq_expression"] = data["expression"]
```

Comment:
I'm still not sure why this is needed. Doesn't the model support this via both `#expression=` and `#miq_expression=` ?

Reply:
This is needed because the setter in the model for `expression=` looks like this: https://github.com/ManageIQ/manageiq/blob/master/app/models/miq_alert.rb#L59-L65
```ruby
    if exp.kind_of?(MiqExpression)
      self.miq_expression = exp
    elsif exp.kind_of?(Hash)
      self.hash_expression = exp
```

And the data we get from the api user is always a "hash" (but not a Hash Expression), because it's parsed json.

The model will only try to create a new MiqExpression from hash data if you explicitly use the `miq_expression` setter, so we need to check if the user provided us an "expression" parameter, and if so *explicitly* say we want a miq_expression and not an hash expression.

Since hash expressions are just ruby hashes, there's no cleaner way to do this - I don't think we'd want the model to attempt parsing hashes to see if they fit the "miq expression" or "hash expression" forms and magically decide which type to use.

Reply:
This is OK for now....can we look at deprecating the old API (i.e. `MiqAlert#expression`) in a follow up PR? Then I think a lot of the complexity of this controller could go away.

Reply:
I'm fine with deprecating the `MiqAlert#expression`, and I can do a follow-up PR.

@moolitayer is there any reason to keep the old `MiqAlert#expression` API around?

Reply:
In the model we would need to keep the `expression` due to the alerts screen.
In the API I agree we should deprecate and then drop.
(deprecate in Gaprindashvili and Drop in H?)